lib/wallet/src/Cardano/Wallet.hs

```diff
@@ -2149,6 +2150,76 @@ readNodeTipStateForTxWrite netLayer = do
             $ ErrNodeNotYetInRecentEra nopp
         Right pp -> pure (pp, timeTranslation)
 
+-- | Wallet-specific wrapped version of 'Write.balanceTx', made for the new tx
+-- workflow with Shelley- and Shared- wallet flavors.
+--
+-- Changes to the change state are not written to the DB.
+balanceTx
+    :: forall s era.
+        ( GenChange s
+        , WalletFlavor s
+        , HasSNetworkId (NetworkOf s)
+        , Write.IsRecentEra era
+        , Including AllFlavors '[ 'ShelleyF, 'SharedF] (FlavorOf s)
+        )
+    => WalletLayer IO s
+    -> Write.PParams era
+    -> TimeTranslation
+    -> PartialTx era
+    -> IO (Write.Tx era)
+balanceTx wrk pp timeTranslation partialTx = do
+    (utxo, wallet, _txs) <- liftIO $ readWalletUTxO wrk
+    -- FIXME: Why are we not using the availableUTxO here?
+    let utxoIndex =
+            Write.constructUTxOIndex $
+            Write.fromWalletUTxO utxo
+
+    let utxoAssumptions = case walletFlavor @s of
+            ShelleyWallet -> AllKeyPaymentCredentials
+            SharedWallet -> AllScriptPaymentCredentialsFrom
+                (Shared.paymentTemplate (getState wallet))
+                (sharedWalletScriptLookup (getState wallet)
+                    . Convert.toWalletAddress)
+
+    let changeState = getState wallet
+    (tx, _changeState') <- throwBalanceTxErr $ Write.balanceTx
+        pp
+        timeTranslation
+        utxoAssumptions
+        utxoIndex
+        (defaultChangeAddressGen argGenChange)
+        changeState
+        partialTx
+
+    return tx
+  where
+    argGenChange :: ArgGenChange s
+    argGenChange = case walletFlavor @s of
+        ShelleyWallet -> delegationAddressS @(NetworkOf s)
+        SharedWallet -> constructAddressFromIx @(NetworkOf s) UtxoInternal
+
+    sharedWalletScriptLookup
+        :: SharedState (NetworkOf s) SharedKey -> Address -> CA.Script KeyHash
+    sharedWalletScriptLookup s addr = case fst (isShared addr s) of
+            Nothing ->
+                error $ "Some inputs selected by coin selection do not belong "
+                <> "to multi-signature wallet"
+            Just (ix,role) ->
+                let template = paymentTemplate s
+                    role' = case role of
+                        UtxoExternal -> CAShelley.UTxOExternal
+                        UtxoInternal -> CAShelley.UTxOInternal
+                        MutableAccount ->
+                            error "role is specified only for payment credential"
+                in replaceCosignersWithVerKeys role' template ix
+
+    throwBalanceTxErr = throwWrappedErr ExceptionBalanceTx
+      where
+        throwOnErr :: (MonadIO m, Exception e) => Either e a -> m a
+        throwOnErr = either (liftIO . throwIO) pure
+
+        throwWrappedErr f e = runExceptT (withExceptT f e) >>= throwOnErr
```

Comment:
I duplicated these from another function, but should probably move to the top level and reuse instead.